### DISCUSSION TITLE
想问一下从umi3.x升级到4.x，有没有通过升级umi相关包的形式，而不是替换成@umijs/max的包的形式

### DISCUSSION THREAD

**Discussion by ZIA-Hans** [umijs/umi#12163] [Q&A] [answered]
已有的项目比较庞大，虽然最好的形式是锁死大版本。但看到umi4.x多了这么多新特性还是想试试的。

**Comment by ZIA-Hans** [comment]
如果按照文档替换@umijs/max的形式的话，原来会有许多地方从umi包引出的方法啥的需要更改。
顺带一提，从升级umi4.x的文档里，已经将umi包替换成@umijs/max包，为啥还会出现从umi包里引出方法的写法，是文档错误吗？
![image](https://github.com/umijs/umi/assets/71582199/42efa121-1ff7-4c22-b28d-5877644e3a28)

**Comment by fz6m** [comment, marked as answer]
关于避免导入 `umi` 都更改为 `@umijs/max` 的问题，有两个解决办法。

第一个方法：在项目根目录新增一个项目级插件 `plugin.ts` ：

```ts
// plugin.ts

import { IApi } from '@umijs/max';

export default (api: IApi) => {
  api.modifyTSConfig(memo => {
    memo.compilerOptions.paths.umi = memo.compilerOptions.paths['@umijs/max']
    return memo
  })
};
```

因为 `umi` 和 `@umijs/max` 导入的东西内容是一样的，只是称呼不一样，所以手动加一个 umi 的 TS 类型提示即可避免迁移 umi 导入名字。

2. 初始化一个 umi 项目，而不是 max 项目，并手动添加使用的插件：

```bash
  pnpm i -D @umijs/plugins
```

```ts
// .umirc.ts

  plugins: [
    '@umijs/plugins/dist/access',
    '@umijs/plugins/dist/antd',
    // ...
  ],
  antd: {},
  access: {},
  // ...
```

max 和 umi 的区别是 max = umi + `@umijs/plugins` ，所以你手动添加进去就变成功能和 max 一样了，max 默认自带的插件参考：[max/preset](https://github.com/umijs/umi/blob/master/packages/max/src/preset.ts) 。

注：在迁移升级的过程中，最好先初始化一个 max / umi 新项目对照着一点点挪，参考文档和网络搜索来探索进行。
